C++ tensor multi-dim indexing: add Slice / TensorIndex

yf225 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #30427 C++ tensor multi-dim indexing: more indexing tests
* #30426 Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h
* #30425 C++ tensor multi-dim indexing: add index() and index_put_() overloads, and simple indexing tests
* **#30424 C++ tensor multi-dim indexing: add Slice / TensorIndex**

`at::indexing::TensorIndex` is used for converting C++ tensor indices such as `{None, "...", Ellipsis, 0, true, {1, None, 2}, torch::tensor({1, 2})}` into its equivalent `std::vector<TensorIndex>`, so that further tensor indexing operations can be performed using the supplied indices.

Differential Revision: [D18695902](https://our.internmc.facebook.com/intern/diff/D18695902)

Commit:
C++ tensor indexing: add Slice / TensorIndex

kostmo commented:
## CircleCI build failures summary
As of commit 3a03b4f:
* **3/3** failures introduced in this PR



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=3a03b4f22e73579c88aa8e46eaf117000378170b):

### 3 failures recognized by patterns:
#### <img src='https://avatars0.githubusercontent.com/ml/7?s=12'/> pytorch_linux_xenial_py2_7_9_test
**Step:** Test
```
Nov 25 22:48:52 [  FAILED  ] TensorIndexingTest.TensorIndex
```
#### <img src='https://avatars0.githubusercontent.com/ml/7?s=12'/> pytorch_linux_xenial_cuda9_cudnn7_py3_multigpu_test
**Step:** Test
```
Nov 25 22:50:36 [  FAILED  ] TensorIndexingTest.TensorIndex
```
#### <img src='https://avatars0.githubusercontent.com/ml/7?s=12'/> pytorch_linux_xenial_py3_6_gcc5_4_test
**Step:** Test
```
Nov 25 22:52:25 [  FAILED  ] TensorIndexingTest.TensorIndex
```


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

ezyang commented:
Sorry, this diff fell off the cart. Looking at it now.

ezyang commented on test/cpp/api/tensor_indexing.cpp:
```diff
@@ -0,0 +1,88 @@
+    ASSERT_TRUE(indices[4].boolean() == true);
+    ASSERT_TRUE(indices[5].is_slice());
+    ASSERT_TRUE(indices[5].slice().start() == 1);
+    ASSERT_TRUE(indices[5].slice().stop() == INDEX_MAX);
```
Interesting. You wrote `None` in the slice specification, but it turns into `INDEX_MAX`?  I guess this is sound because you can never out of bound slice ranges in PYthon.

ezyang commented on test/cpp/api/tensor_indexing.cpp:
```diff
@@ -0,0 +1,88 @@
+
+  // NOTE: Some compilers such as Clang 5 and MSVC always treat `TensorIndex({1})` the same as
+  // `TensorIndex(1)`. Therefore, we should only run this test on compilers that treat
+  // `TensorIndex({1})` as `TensorIndex(std::initializer_list<c10::optional<int64_t>>({1}))`.
```
What does the C++ standard say should happen in this case?

ezyang commented on aten/src/ATen/native/TensorIndexing.h:
```diff
@@ -0,0 +1,110 @@
+  Slice();
+  Slice(int64_t start, int64_t stop, int64_t step);
+
+  const int64_t& start() const;
```
Why do these return const references?

ezyang commented on aten/src/ATen/native/TensorIndexing.h:
```diff
@@ -0,0 +1,110 @@
+
+  // Case 3: Integer value
+  TensorIndex(int64_t integer);
+  TensorIndex(int integer);
```
Why do you also need an `int` overload here?

yf225 replied:
We need the `TensorIndex(int)` overload because otherwise e.g. `TensorIndex(5)` would be ambiguous:
```
../test/cpp/api/tensor_indexing.cpp:58:85: error: conversion from ‘int’ to ‘at::indexing::TensorIndex’ is ambiguous
     std::vector<TensorIndex> indices = {None, "...", Ellipsis, 0, true, {1, None, 2}};
                                                                                     ^
In file included from ../test/cpp/api/tensor_indexing.cpp:5:0:
../aten/src/ATen/native/TensorIndexing.h:70:3: note: candidate: at::indexing::TensorIndex::TensorIndex(int64_t)
   TensorIndex(int64_t integer);
   ^
../aten/src/ATen/native/TensorIndexing.h:67:3: note: candidate: at::indexing::TensorIndex::TensorIndex(const char*)
   TensorIndex(const char *str);
   ^
```

ezyang commented on aten/src/ATen/native/TensorIndexing.h:
```diff
@@ -0,0 +1,110 @@
+  TensorIndex(int integer);
+
+  // Case 4: Boolean value
+  TensorIndex(bool boolean);
```
It might be a good idea to make sure that you don't do implicit conversions into boolean, since this can lead to some nasty bugs; e.g., your current constructor accepts a `std::shared_ptr<Tensor>`. You can do that with a template. (cc @zdevito)

ezyang commented on aten/src/ATen/native/TensorIndexing.cpp:
```diff
@@ -0,0 +1,159 @@
+}
+
+std::ostream& operator<<(std::ostream& stream, const Slice& slice) {
+  stream << "{" << slice.start() << ", " << slice.stop() << ", " << slice.step() << "}";
```
I am wondering why you stringify slices as the initializer list, as opposed to the Python syntax (e.g., `2::3` versus `{2, None, 3}`. It seems like it would be helpful to be able to report a slice in the Python syntax.

ezyang commented on aten/src/ATen/native/TensorIndexing.cpp:
```diff
@@ -0,0 +1,159 @@
+    if (step == 0) {
+      TORCH_CHECK(false, "slice step cannot be zero");
+    }
+    // Here *step might be -INDEX_MAX-1; in this case we replace it
```
you mean step?

ezyang reviewed (approved):
Looks reasonable.

yf225 commented on aten/src/ATen/native/TensorIndexing.h:
```diff
@@ -0,0 +1,110 @@
+#pragma once
+
+#include <c10/util/Optional.h>
+#include <ATen/core/TensorBody.h>
```
TODO: have a top level comment on this header explaining the general strategy here.

yf225 commented on aten/src/ATen/native/TensorIndexing.h:
```diff
@@ -0,0 +1,110 @@
+
+constexpr c10::nullopt_t None{c10::nullopt_t::init()};
+
+struct CAFFE2_API EllipsisIndexType { EllipsisIndexType() {} };
```
TODO: make this `final`

yf225 commented on aten/src/ATen/native/TensorIndexing.h:
```diff
@@ -0,0 +1,110 @@
+struct CAFFE2_API EllipsisIndexType { EllipsisIndexType() {} };
+CAFFE2_API extern const EllipsisIndexType Ellipsis;
+
+struct CAFFE2_API Slice {
```
TODO: make this `final`

yf225 commented on aten/src/ATen/native/TensorIndexing.h:
```diff
@@ -0,0 +1,110 @@
+// `:3:2`                  | `{None, 3, 2}`
+// `1:3:2`                 | `{1, 3, 2}`
+// `torch.tensor([1, 2])`) | `torch::tensor({1, 2})`
+struct CAFFE2_API TensorIndex {
```
TODO: make this `final`

Commit:
Update on "C++ tensor indexing: add Slice / TensorIndex"

C++ tensor indexing: add Slice / TensorIndex

gh-metadata: pytorch pytorch 30424 gh/yf225/86/head

Commit:
Update on "C++ tensor indexing: add Slice / TensorIndex"

C++ tensor indexing: add Slice / TensorIndex

gh-metadata: pytorch pytorch 30424 gh/yf225/86/head

Commit:
Update on "C++ tensor indexing: add Slice / TensorIndex"

C++ tensor indexing: add Slice / TensorIndex

gh-metadata: pytorch pytorch 30424 gh/yf225/86/head

Commit:
Update on "C++ tensor indexing: add Slice / TensorIndex"

C++ tensor indexing: add Slice / TensorIndex

gh-metadata: pytorch pytorch 30424 gh/yf225/86/head

Commit:
Update on "C++ tensor indexing: add Slice / TensorIndex"

C++ tensor indexing: add Slice / TensorIndex

gh-metadata: pytorch pytorch 30424 gh/yf225/86/head

Commit:
Update on "C++ tensor indexing: add Slice / TensorIndex"

C++ tensor indexing: add Slice / TensorIndex

gh-metadata: pytorch pytorch 30424 gh/yf225/86/head

Commit:
Update on "C++ tensor indexing: add Slice / TensorIndex"

C++ tensor indexing: add Slice / TensorIndex

gh-metadata: pytorch pytorch 30424 gh/yf225/86/head

Commit:
Update on "C++ tensor indexing: add Slice / TensorIndex"

C++ tensor indexing: add Slice / TensorIndex

gh-metadata: pytorch pytorch 30424 gh/yf225/86/head
